src/celeritas/phys/RootPrimaryGenerator.cc

```diff
@@ -0,0 +1,121 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/phys/RootPrimaryGenerator.cc
+//---------------------------------------------------------------------------//
+#include "RootPrimaryGenerator.hh"
+
+#include <TFile.h>
+#include <TLeaf.h>
+#include <TTree.h>
+
+#include "corecel/Macros.hh"
+#include "corecel/cont/Range.hh"
+#include "corecel/io/Logger.hh"
+#include "celeritas/ext/ScopedRootErrorHandler.hh"
+#include "celeritas/phys/ParticleParams.hh"
+#include "celeritas/phys/Primary.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from input file, number of events, and primaries to be sampled per
+ * event.
+ */
+RootPrimaryGenerator::RootPrimaryGenerator(std::string const& filename,
+                                           SPConstParticles params,
+                                           size_type num_events,
+                                           size_type primaries_per_event,
+                                           unsigned int seed)
+    : params_(std::move(params))
+    , num_events_{num_events}
+    , primaries_per_event_(primaries_per_event)
+{
+    CELER_EXPECT(!filename.empty());
+    CELER_VALIDATE(num_events > 0 && primaries_per_event > 0,
+                   << "'num_events' and 'primaries_per_event options in input "
+                      "file must be set and positive");
+
+    ScopedRootErrorHandler scoped_root_error;
+    tfile_.reset(TFile::Open(filename.c_str(), "read"));
+    CELER_ENSURE(tfile_->IsOpen());
+
+    ttree_.reset(tfile_->Get<TTree>("primaries"));
+    CELER_ENSURE(ttree_);
+    CELER_VALIDATE(ttree_->GetEntries() > 0,
+                   << "TTree `primaries` in '" << tfile_->GetName()
+                   << "' has zero entries");
+    scoped_root_error.throw_if_errors();
+
+    entry_selector_ = std::uniform_int_distribution<>(0, primaries_per_event);
```

Comment:
name function-like class instances like you would name a function
```suggestion
    select_entry_ = std::uniform_int_distribution<>(0, primaries_per_event);
```